### PR TITLE
worker: implement `MessagePort` and `MessageChannel`

### PR DESCRIPTION
**Pull request by addaleax**
Implement `MessagePort` and `MessageChannel` along the lines of
the DOM classes of the same names. `MessagePort`s initially
support transferring only `ArrayBuffer`s.

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added (not yet, but see #58)
- [x] commit message follows [commit guidelines](https://github.com/ayojs/ayo/blob/latest/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)

worker

### CODE REVIEW

File: src/env.h [hunk 1]
```diff
@@ -92,6 +92,7 @@ class ModuleWrap;
   V(decorated_private_symbol, "node:decorated")                               \
   V(npn_buffer_private_symbol, "node:npnBuffer")                              \
   V(processed_private_symbol, "node:processed")                               \
+  V(sab_lifetimepartner_symbol, "node:sharedArrayBufferLiftimePartner")       \
```
Comment (hybrist) [inline src/env.h]: typo? `node:sharedArrayBufferLiftimePartner -> node:sharedArrayBufferLifetimePartner`

Comment (addaleax) [reply src/env.h]: a) yes, thank you and b) it wasn’t even meant to be in this commit :smile: I’ve fixed it in #58, though

**Comment by hybrist** [comment]
References:

* [`MessagePort`](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort)
* [`MessageChannel`](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel)
* [HTML spec](https://html.spec.whatwg.org/multipage/web-messaging.html#channel-messaging)

**Comment by hybrist** [comment]
I assume "along the lines" is mostly referring to the differences in event listener interfaces (`EventEmitter` vs. `EventTarget`)? Any other intended limitations / deviations?

File: lib/internal/worker.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  debug(`[${process.threadId}] received message`, flag, payload);
+  // Emit the flag and deserialized object to userland.
+  if (flag === 0 || flag === undefined)
+    this.emit('message', payload);
+  else
+    this.emit('flaggedMessage', flag, payload);
+}
+
+Object.defineProperty(MessagePort.prototype, 'onmessage', {
+  enumerable: false,
```
Comment (TimothyGu) [inline lib/internal/worker.js]: Make it enumerable for consistency.

File: src/env.h [hunk 3]
```diff
@@ -317,6 +322,7 @@ class ModuleWrap;
   V(promise_wrap_template, v8::ObjectTemplate)                                \
   V(push_values_to_array_function, v8::Function)                              \
   V(randombytes_constructor_template, v8::ObjectTemplate)                     \
+  V(sab_lifetimepartner_constructor_template, v8::FunctionTemplate)           \
```
Comment (TimothyGu) [inline src/env.h]: Should this be in this commit?

File: src/node_messaging.cc [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+using v8::MaybeLocal;
+using v8::Nothing;
+using v8::Object;
+using v8::String;
+using v8::Value;
+using v8::ValueDeserializer;
+using v8::ValueSerializer;
+
+namespace node {
+namespace worker {
```
Comment (TimothyGu) [inline src/node_messaging.cc:34]: Make it `namespace messaging` for consistency with the file name?

**Review by TimothyGu — COMMENTED** [review]
More review coming.

**Comment by TimothyGu** [comment]
@jkrems 

> I assume "along the lines" is mostly referring to the differences in event listener interfaces (`EventEmitter` vs. `EventTarget`)?

There's that, and also some additional APIs like `oninit`, `stop()`, `ref()`, `unref()`, and `hasRef()`.

File: src/node_messaging.cc [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+void MessagePort::Entangle(MessagePort* a, MessagePortData* b) {
+  MessagePortData::Entangle(a->data_.get(), b);
+}
+
+MaybeLocal<Function> GetMessagePortConstructor(
+    Environment* env, Local<Context> context) {
+  // Factor generating the MessagePort JS constructor into its own piece
+  // of code, because it is needed early on in the child environment setup.
+  Local<FunctionTemplate> templ;
+  templ = env->message_port_constructor_template();
```
Comment (TimothyGu) [inline src/node_messaging.cc]: Merge these two lines?

File: lib/internal/worker.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    Object.defineProperty(this, {
+      writable: true,
+      enumerable: true,
+      configurable: true,
+      value
+    });
+    this.ref();
+    this.start();
+  }
+});
```
Comment (Qard) [inline lib/internal/worker.js:38]: Changing `onmessage` will stop the `message` and `flaggedMessage` events from being emitted. Is that intended behaviour?

**Review by Qard — COMMENTED** [review]
Other than one question, it all LGTM.

File: lib/internal/worker.js (hunk 2 above)
Comment (addaleax) [reply lib/internal/worker.js]: Hm – the idea here was consistency with ES6 class methods, but I guess you’re right, this should match the native ones

File: src/env.h (hunk 3 above)
Comment (addaleax) [reply src/env.h]: nope :)

File: src/node_messaging.cc (hunk 4 above)
Comment (addaleax) [reply src/node_messaging.cc:34]: well … it’s still conceptually part of the `worker` code, isn’t it?

File: src/node_messaging.cc (hunk 5 above)
Comment (addaleax) [reply src/node_messaging.cc]: done :)

File: lib/internal/worker.js (hunk 6 above)
Comment (addaleax) [reply lib/internal/worker.js:38]: Yeah, I would say so. This is only there to match the Web spec more closely, but in that case you don’t want the EventEmitter interface anyways I guess.

**Comment by hybrist** [comment]
@TimothyGu For context - I was trying to figure out how hard it would be to write portable code against this, so I wasn't worried too much about additional properties and more about meaningful API overlap. I assume the absence of `.addEventListener` in node makes this a bit annoying. But short of extending `EventEmitter` itself I don't think that can be fixed. :(

**Comment by addaleax** [comment]
@jkrems Hm, yes. We could add missing methods to the `MessagePort` prototype itself, if we want?

**Comment by hybrist** [comment]
I'm somewhat partial towards "make it a clean superset of the existing/spec browser API" but I'm not sure how realistic that is (and how much value it would provide in practice). Adding `.addEventListener` to these classes might be a neat cheap win in that regard.

**Comment by addaleax** [comment]
@jkrems Do you want to try & take a stab at this yourself? The trickiest thing I guess would be the interaction of all 3 event emitter systems (`.on()`, `.addEventListener()`, raw `.onmessage = foo`)…

**Comment by Qard** [comment]
The missing bits could probably also just be added later.

**Review by Qard — APPROVED** [review]

**Comment by addaleax** [comment]
Landed in f65bb3dee5b41932295cd1e77839c63ec0072bfa

**Pull request closed without merging**